core/src/main/java/org/apache/calcite/rel/type/RelDataTypeFactoryImpl.java

```diff
@@ -410,6 +410,31 @@ private RelDataType copyRecordType(
     return canonize(newType);
   }
 
+  @Override public RelDataType enforceTypeWithNullability(
+      final RelDataType type,
+      final boolean nullable) {
+    requireNonNull(type, "type");
+    RelDataType newType;
+    if (type.isNullable() == nullable) {
+      newType = type;
+    } else if (type instanceof RelRecordType) {
+      return createStructType(type.getStructKind(),
+          new AbstractList<RelDataType>() {
+            @Override public RelDataType get(int index) {
+              return type.getFieldList().get(index).getType();
+            }
+
+            @Override public int size() {
+              return type.getFieldCount();
+            }
+          },
+          type.getFieldNames(), nullable);
+    } else {
```

Comment:
Aren't there other non simple types like maps (`MultisetSqlType`)/arrays (`ArraySqlType`)? Is it enough in this form?

EDIT: I have just seen you have done that for the `SqlTypeFactoryImpl`, why here it is not needed?

Reply:
As far as I can tell the other types are handled correctly by the factory.
Only Record types are anomalous.
This class (RelDataTypeFactoryImpl) is an abstract base class out of which SqlTypeFactoryImpl is derived. In the end, the SqlTypeFactoryImpl method gets called.

Reply:
Thanks for the clarification Mihai, I have no more questions, hence I will approve the PR, the rest is minor and I will leave it to your judgement

Reply:
I will document it anyway.

Reply:
I guess so, because implementers will have to add support for it

Reply:
I will add this to history.md in the section on breaking changes:
```
*RelDataTypeFactory interface*.  The fix for [<a
href="https://issues.apache.org/jira/browse/CALCITE-6764">CALCITE-6764</a>]
introduces a new method
`RelDataTypeFactory#enforceTypeWithNullability` in the existing
`RelDataTypeFactory` interface.  The behavior of the new function is
similar to the existing API `createTypeWithNullability`; however, the
existing implementations of the `createTypeWithNullability` API cannot
create nullable record (`ROW`) types.  Nullable record types are
legitimate in several SQL dialects.
```